From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 22 Apr 2026 01:56:47 +0800
Subject: [PATCH] Add configurable region format framework & linear v2 region
 format support

Co-authored by: MrHua269 <wangxyper@163.com>
Luminol (https://github.com/LuminolMC/Luminol)
Licensed under: MIT (https://opensource.org/licenses/MIT)

diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java
index a814512fcfb85312474ae2c2c21443843bf57831..2e084a5b28cbe4737f48c25e10af589213525362 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java
@@ -8,9 +8,9 @@ public interface ChunkSystemRegionFileStorage {
 
     public boolean moonrise$doesRegionFileNotExistNoIO(final int chunkX, final int chunkZ);
 
-    public RegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ);
+    public abomination.IRegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ); // Luminol - Configurable region file format
 
-    public RegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException;
+    public abomination.IRegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException; // Luminol - Configurable region file format
 
     public MoonriseRegionFileIO.RegionDataController.WriteData moonrise$startWrite(
             final int chunkX, final int chunkZ, final CompoundTag compound
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
index 3bb9b58ee97464687e348e23b99159226415c267..fc264727141c3897e6c06819b9713d4f6395baf0 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
@@ -1273,7 +1273,7 @@ public final class MoonriseRegionFileIO {
                 this.regionDataController.finishWrite(this.chunkX, this.chunkZ, writeData);
                 // Paper start - flush regionfiles on save
                 if (this.world.paperConfig().chunks.flushRegionsOnSave) {
-                    final RegionFile regionFile = this.regionDataController.getCache().moonrise$getRegionFileIfLoaded(this.chunkX, this.chunkZ);
+                    final abomination.IRegionFile regionFile = this.regionDataController.getCache().moonrise$getRegionFileIfLoaded(this.chunkX, this.chunkZ); // Luminol - Add configurable region file
                     if (regionFile != null) {
                         regionFile.flush();
                     } // else: evicted from cache, which should have called flush
@@ -1489,7 +1489,7 @@ public final class MoonriseRegionFileIO {
 
         public static interface IORunnable {
 
-            public void run(final RegionFile regionFile) throws IOException;
+            public void run(final abomination.IRegionFile regionFile) throws IOException; // Luminol - Configurable region file format
 
         }
     }
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/storage/ChunkSystemChunkBuffer.java b/ca/spottedleaf/moonrise/patches/chunk_system/storage/ChunkSystemChunkBuffer.java
index 51c126735ace8fdde89ad97b5cab62f244212db0..c7d4d944eb198ac53a3eeae717a25c7d5815c8c1 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/storage/ChunkSystemChunkBuffer.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/storage/ChunkSystemChunkBuffer.java
@@ -8,5 +8,5 @@ public interface ChunkSystemChunkBuffer {
 
     public void moonrise$setWriteOnClose(final boolean value);
 
-    public void moonrise$write(final RegionFile regionFile) throws IOException;
+    public void moonrise$write(final abomination.IRegionFile regionFile) throws IOException; // Luminol - Configurable region file format
 }
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index a9104198a8a4ccfdde705a1506ac0606bee5ea1e..b99bb8e68e45b04028a390e203ca73eba56e8c0c 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1008,10 +1008,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         if (flush) {
             for (ServerLevel level : this.getAllLevels()) {
                 String storageName = level.getChunkSource().chunkMap.getStorageName();
-                LOGGER.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", LEGACY_WORLD_NAMES_FOR_REALMS_LOG.getOrDefault(storageName, storageName));
+                LOGGER.info("ThreadedChunkStorage ({}): All chunks are saved", LEGACY_WORLD_NAMES_FOR_REALMS_LOG.getOrDefault(storageName, storageName)); // Luminol - configurable region format
             }
 
-            LOGGER.info("ThreadedAnvilChunkStorage: All dimensions are saved");
+            LOGGER.info("ThreadedChunkStorage: All dimensions are saved"); // Luminol - configurable region format
         }
 
         return result;
diff --git a/net/minecraft/util/worldupdate/FileToUpgrade.java b/net/minecraft/util/worldupdate/FileToUpgrade.java
index a7f2cfa9277c898038f6b9e0a3401db51012b64c..db4dba8849eb27f8f84acc3297e342da4700d839 100644
--- a/net/minecraft/util/worldupdate/FileToUpgrade.java
+++ b/net/minecraft/util/worldupdate/FileToUpgrade.java
@@ -4,5 +4,5 @@ import java.util.List;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.chunk.storage.RegionFile;
 
-public record FileToUpgrade(RegionFile file, List<ChunkPos> chunksToUpgrade) {
+public record FileToUpgrade(abomination.IRegionFile file, List<ChunkPos> chunksToUpgrade) { // Luminol - Configurable region file format
 }
diff --git a/net/minecraft/util/worldupdate/RegionStorageUpgrader.java b/net/minecraft/util/worldupdate/RegionStorageUpgrader.java
index c23fc1254ae54ca3996aec8a4a7ffe66c1e8d8f9..8305a5fc445108f8bf9ee8c767fe96e7ad49cb26 100644
--- a/net/minecraft/util/worldupdate/RegionStorageUpgrader.java
+++ b/net/minecraft/util/worldupdate/RegionStorageUpgrader.java
@@ -36,7 +36,7 @@ import org.slf4j.Logger;
 public class RegionStorageUpgrader {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final String NEW_DIRECTORY_PREFIX = "new_";
-    private static final Pattern REGEX = Pattern.compile("^r\\.(-?[0-9]+)\\.(-?[0-9]+)\\.mca$");
+    private static final Pattern REGEX = Pattern.compile("^r\\.(-?[0-9]+)\\.(-?[0-9]+)\\\\"+ net.minecraft.world.level.chunk.storage.RegionFileStorage.getExtensionName() +"$"); // Luminol - Configurable region file format
     private final DataFixer dataFixer;
     private final UpgradeProgress upgradeProgress;
     private final String type;
@@ -160,7 +160,7 @@ public class RegionStorageUpgrader {
     }
 
     private static List<FileToUpgrade> getAllChunkPositions(final RegionStorageInfo info, final Path regionFolder) {
-        File[] files = regionFolder.toFile().listFiles((dir, name) -> name.endsWith(".mca"));
+        File[] files = regionFolder.toFile().listFiles((dir, name) -> name.endsWith(net.minecraft.world.level.chunk.storage.RegionFileStorage.getExtensionName())); // Luminol - Configurable region file format
         if (files == null) {
             return List.of();
         } else {
@@ -173,7 +173,7 @@ public class RegionStorageUpgrader {
                     int zOffset = Integer.parseInt(regex.group(2)) << 5;
                     List<ChunkPos> chunkPositions = Lists.newArrayList();
 
-                    try (RegionFile regionSource = new RegionFile(info, regionFile.toPath(), regionFolder, true)) {
+                    try (abomination.IRegionFile regionSource = net.minecraft.world.level.chunk.storage.RegionFileStorage.createNew(info, regionFile.toPath(), regionFolder, true)) { // Luminol - Configurable region file format
                         for (int x = 0; x < 32; x++) {
                             for (int z = 0; z < 32; z++) {
                                 ChunkPos pos = new ChunkPos(x + xOffset, z + zOffset);
@@ -251,7 +251,7 @@ public class RegionStorageUpgrader {
         return storage.upgradeChunkTag(chunkTag, this.defaultVersion, this.dataFixContextTag, targetVersion);
     }
 
-    private void onFileFinished(final RegionFile regionFile) {
+    private void onFileFinished(final abomination.IRegionFile regionFile) { // Luminol - Configurable region file format
         if (this.recreateRegionFiles) {
             if (this.previousWriteFuture != null) {
                 this.previousWriteFuture.join();
diff --git a/net/minecraft/world/level/chunk/storage/RegionFile.java b/net/minecraft/world/level/chunk/storage/RegionFile.java
index e077797cefffcb9db627cab12c7d88245e5baa69..35a49cc3c1ad01623c05208480483565faa4b881 100644
--- a/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -23,7 +23,7 @@ import net.minecraft.world.level.ChunkPos;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.storage.ChunkSystemRegionFile { // Paper - rewrite chunk system
+public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.storage.ChunkSystemRegionFile, abomination.IRegionFile { // Paper - rewrite chunk system // Luminol - Configurable region file format
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final int MAX_CHUNK_SIZE = 500 * 1024 * 1024; // Paper - don't write garbage data to disk if writing serialization fails
     private static final int SECTOR_BYTES = 4096;
@@ -790,7 +790,7 @@ public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patche
         }
     }
 
-    protected synchronized void write(final ChunkPos pos, final ByteBuffer data) throws IOException {
+    public synchronized void write(final ChunkPos pos, final ByteBuffer data) throws IOException { // Luminol - make public
         int offsetIndex = getOffsetIndex(pos);
         int offset = this.offsets.get(offsetIndex);
         int sectorNumber = getSectorNumber(offset);
@@ -856,7 +856,7 @@ public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patche
         this.file.write(this.header, 0L);
     }
 
-    private int getOffset(final ChunkPos pos) {
+    public int getOffset(final ChunkPos pos) { // Luminol - make public
         return this.offsets.get(getOffsetIndex(pos));
     }
 
@@ -908,8 +908,8 @@ public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patche
         }
 
         @Override
-        public final void moonrise$write(final RegionFile regionFile) throws IOException {
-            regionFile.write(this.pos, ByteBuffer.wrap(this.buf, 0, this.count));
+        public final void moonrise$write(final abomination.IRegionFile regionFile) throws IOException { // Luminol - Configurable region file format
+            ((RegionFile)regionFile).write(this.pos, ByteBuffer.wrap(this.buf, 0, this.count));
         }
         // Paper end - rewrite chunk system
 
diff --git a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
index eae4963b6d18c5f92871477b93f81c2c79331cd9..336d636b7e906fd4bdfa3a658fcd844628d71c88 100644
--- a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
@@ -19,10 +19,22 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     private static final org.slf4j.Logger LOGGER = com.mojang.logging.LogUtils.getLogger(); // Paper
     public static final String ANVIL_EXTENSION = ".mca";
     private static final int MAX_CACHE_SIZE = 256;
-    public final Long2ObjectLinkedOpenHashMap<RegionFile> regionCache = new Long2ObjectLinkedOpenHashMap<>();
+    public final Long2ObjectLinkedOpenHashMap<abomination.IRegionFile> regionCache = new Long2ObjectLinkedOpenHashMap<>(); // Luminol - Configurable region file format
     private final RegionStorageInfo info;
     private final Path folder;
     private final boolean sync;
+    private final boolean isLinear; // Luminol - Configurable region file format
+
+    public static String getExtensionName() { // Luminol - Configurable region file format
+        return dev.bacteriawa.mint.config.modules.misc.RegionFormatConfig.regionFormat == abomination.EnumRegionFormat.LINEAR ? ".linear" : ANVIL_EXTENSION;
+    }
+
+    public static abomination.IRegionFile createNew(RegionStorageInfo info, Path path, Path externalFileDir, boolean sync) throws IOException { // Luminol - Configurable region file format
+        if (dev.bacteriawa.mint.config.modules.misc.RegionFormatConfig.regionFormat == abomination.EnumRegionFormat.LINEAR) {
+            return new abomination.LinearRegionFile(path, io.papermc.paper.configuration.GlobalConfiguration.get().misc.linearRegionCompressionLevel);
+        }
+        return new RegionFile(info, path, externalFileDir, sync);
+    }
 
     // Paper start - recalculate region file headers
     private final boolean isChunkData;
@@ -30,7 +42,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     @Nullable
     public static ChunkPos getRegionFileCoordinates(Path file) {
         String fileName = file.getFileName().toString();
-        if (!fileName.startsWith("r.") || !fileName.endsWith(".mca")) {
+        if (!fileName.startsWith("r.") || !fileName.endsWith(getExtensionName())) { // Luminol - Configurable region file format
             return null;
         }
 
@@ -55,7 +67,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     private static final int MAX_NON_EXISTING_CACHE = 1024 * 4;
     private final it.unimi.dsi.fastutil.longs.LongLinkedOpenHashSet nonExistingRegionFiles = new it.unimi.dsi.fastutil.longs.LongLinkedOpenHashSet();
     private static String getRegionFileName(final int chunkX, final int chunkZ) {
-        return "r." + (chunkX >> REGION_SHIFT) + "." + (chunkZ >> REGION_SHIFT) + ".mca";
+        return "r." + (chunkX >> REGION_SHIFT) + "." + (chunkZ >> REGION_SHIFT) + getExtensionName(); // Luminol - Configurable region file format
     }
 
     private boolean doesRegionFilePossiblyExist(final long position) {
@@ -90,15 +102,15 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     }
 
     @Override
-    public synchronized final RegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ) {
+    public synchronized final abomination.IRegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ) { // Luminol - Configurable region file format
         return this.regionCache.getAndMoveToFirst(ChunkPos.pack(chunkX >> REGION_SHIFT, chunkZ >> REGION_SHIFT));
     }
 
     @Override
-    public synchronized final RegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException {
+    public synchronized final abomination.IRegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException { // Luminol - Configurable region file format
         final long key = ChunkPos.pack(chunkX >> REGION_SHIFT, chunkZ >> REGION_SHIFT);
 
-        RegionFile ret = this.regionCache.getAndMoveToFirst(key);
+        abomination.IRegionFile ret = this.regionCache.getAndMoveToFirst(key); // Luminol - Configurable region file format
         if (ret != null) {
             return ret;
         }
@@ -124,7 +136,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
         FileUtil.createDirectoriesSafe(this.folder);
 
-        ret = new RegionFile(this.info, regionPath, this.folder, this.sync);
+        ret = createNew(this.info, regionPath, this.folder, this.sync); // Luminol - Configurable region file format
 
         this.regionCache.putAndMoveToFirst(key, ret);
 
@@ -143,7 +155,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         }
 
         final ChunkPos pos = new ChunkPos(chunkX, chunkZ);
-        final RegionFile regionFile = this.getRegionFile(pos);
+        final abomination.IRegionFile regionFile = this.getRegionFile(pos); // Luminol - Configurable region file format
 
         // note: not required to keep regionfile loaded after this call, as the write param takes a regionfile as input
         // (and, the regionfile parameter is unused for writing until the write call)
@@ -159,7 +171,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         } catch (final RegionFileSizeException ex) {
             // note: it's OK if close() is called, as close() here will not issue a write to the RegionFile
             // see startWrite
-            final int maxSize = RegionFile.MAX_CHUNK_SIZE / (1024 * 1024);
+            final int maxSize = (regionFile instanceof RegionFile ? RegionFile.MAX_CHUNK_SIZE : abomination.LinearRegionFile.MAX_CHUNK_SIZE) / (1024 * 1024); // Luminol - Configurable region file format
             LOGGER.error("Chunk at (" + chunkX + "," + chunkZ + ") in regionfile '" + regionFile.getPath().toString() + "' exceeds max size of " + maxSize + "MiB, it has been deleted from disk.");
             return new ca.spottedleaf.moonrise.patches.chunk_system.io.MoonriseRegionFileIO.RegionDataController.WriteData(
                 compound, ca.spottedleaf.moonrise.patches.chunk_system.io.MoonriseRegionFileIO.RegionDataController.WriteData.WriteResult.DELETE,
@@ -177,7 +189,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     ) throws IOException {
         final ChunkPos pos = new ChunkPos(chunkX, chunkZ);
         if (writeData.result() == ca.spottedleaf.moonrise.patches.chunk_system.io.MoonriseRegionFileIO.RegionDataController.WriteData.WriteResult.DELETE) {
-            final RegionFile regionFile = this.moonrise$getRegionFileIfExists(chunkX, chunkZ);
+            final abomination.IRegionFile regionFile = this.moonrise$getRegionFileIfExists(chunkX, chunkZ); // Luminol - Configurable region file format
             if (regionFile != null) {
                 regionFile.clear(pos);
             } // else: didn't exist
@@ -192,7 +204,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     public final ca.spottedleaf.moonrise.patches.chunk_system.io.MoonriseRegionFileIO.RegionDataController.ReadData moonrise$readData(
         final int chunkX, final int chunkZ
     ) throws IOException {
-        final RegionFile regionFile = this.moonrise$getRegionFileIfExists(chunkX, chunkZ);
+        final abomination.IRegionFile regionFile = this.moonrise$getRegionFileIfExists(chunkX, chunkZ); // Luminol - Configurable region file format
 
         final DataInputStream input = regionFile == null ? null : regionFile.getChunkDataInputStream(new ChunkPos(chunkX, chunkZ));
 
@@ -237,15 +249,15 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
             final ChunkPos pos = new ChunkPos(chunkX, chunkZ);
             final ChunkPos headerChunkPos = SerializableChunkData.getChunkCoordinate(ret);
-            final RegionFile regionFile = this.getRegionFile(pos);
+            final abomination.IRegionFile regionFile = this.getRegionFile(pos); // Luminol - Configurable region file format
 
-            if (regionFile.getRecalculateCount() != readData.recalculateCount()) {
+            if (regionFile instanceof RegionFile mcaFile && mcaFile.getRecalculateCount() != readData.recalculateCount()) { // Luminol - Configurable region file format
                 return null;
             }
 
             if (!headerChunkPos.equals(pos)) {
                 LOGGER.error("Attempting to read chunk data at " + pos + " but got chunk data for " + headerChunkPos + " instead! Attempting regionfile recalculation " + regionFile.getPath().toAbsolutePath());
-                if (regionFile.recalculateHeader()) {
+                if (regionFile instanceof RegionFile mcaFile && mcaFile.recalculateHeader()) { // Luminol - Configurable region file format
                     return null;
                 }
 
@@ -261,7 +273,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     }
     // Paper end - rewrite chunk system
     // Paper start - rewrite chunk system
-    public RegionFile getRegionFile(ChunkPos chunkcoordintpair) throws IOException {
+    public abomination.IRegionFile getRegionFile(ChunkPos chunkcoordintpair) throws IOException { // Luminol - Configurable region file format
         return this.getRegionFile(chunkcoordintpair, false);
     }
     // Paper end - rewrite chunk system
@@ -271,9 +283,10 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         this.sync = sync;
         this.info = info;
         this.isChunkData = info.dfuType()[0] == net.minecraft.util.datafix.DataFixTypes.CHUNK; // Paper - recalculate region file headers
+        this.isLinear = dev.bacteriawa.mint.config.modules.misc.RegionFormatConfig.regionFormat == abomination.EnumRegionFormat.LINEAR; // Luminol - Configurable region file format
     }
 
-    @org.jetbrains.annotations.Contract("_, false -> !null") private @Nullable RegionFile getRegionFile(final ChunkPos pos, boolean existingOnly) throws IOException { // CraftBukkit
+    @org.jetbrains.annotations.Contract("_, false -> !null") private @Nullable abomination.IRegionFile getRegionFile(final ChunkPos pos, boolean existingOnly) throws IOException { // CraftBukkit // Luminol - Configurable region file format
         // Paper start - rewrite chunk system
         if (existingOnly) {
             return this.moonrise$getRegionFileIfExists(pos.x(), pos.z());
@@ -346,15 +359,15 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
     public @Nullable CompoundTag read(final ChunkPos pos) throws IOException {
         // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
-        RegionFile region = this.getRegionFile(pos, true);
+        abomination.IRegionFile region = this.getRegionFile(pos, true); // Luminol - Configurable region file format
         if (region == null) {
             return null;
         }
         // CraftBukkit end
         // Paper start
-        if (region.isOversized(pos.x(), pos.z())) {
+        if (region instanceof RegionFile mcaFile && mcaFile.isOversized(pos.x(), pos.z())) { // Luminol - Configurable region file format
             printOversizedLog("Loading Oversized Chunk!", region.getPath(), pos.x(), pos.z());
-            return readOversizedChunk(region, pos);
+            return readOversizedChunk(mcaFile, pos); // Luminol - Configurable region file format
         }
         // Paper end
 
@@ -370,7 +383,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
                 ChunkPos headerChunkPos = SerializableChunkData.getChunkCoordinate(var4);
                 if (!headerChunkPos.equals(pos)) {
                     net.minecraft.server.MinecraftServer.LOGGER.error("Attempting to read chunk data at " + pos + " but got chunk data for " + headerChunkPos + " instead! Attempting regionfile recalculation for regionfile " + region.getPath().toAbsolutePath());
-                    if (region.recalculateHeader()) {
+                    if (region instanceof RegionFile mcaFile && mcaFile.recalculateHeader()) { // Luminol - Configurable region file format
                         return this.read(pos);
                     }
                     net.minecraft.server.MinecraftServer.LOGGER.error("Can't recalculate regionfile header, regenerating chunk " + pos + " for " + region.getPath().toAbsolutePath());
@@ -385,11 +398,10 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
     public void scanChunk(final ChunkPos pos, final StreamTagVisitor scanner) throws IOException {
         // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
-        RegionFile region = this.getRegionFile(pos, true);
+        abomination.IRegionFile region = this.getRegionFile(pos, true); // Luminol - Configurable region file format
         if (region == null) {
             return;
         }
-        // CraftBukkit end
 
         try (DataInputStream regionChunkInputStream = region.getChunkDataInputStream(pos)) {
             if (regionChunkInputStream != null) {
@@ -400,7 +412,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
     public void write(final ChunkPos pos, final @Nullable CompoundTag value) throws IOException { // Paper - rewrite chunk system - public
         if (!SharedConstants.DEBUG_DONT_SAVE_WORLD) {
-            RegionFile region = this.getRegionFile(pos, value == null); // CraftBukkit // Paper - rewrite chunk system
+            abomination.IRegionFile region = this.getRegionFile(pos, value == null); // CraftBukkit // Paper - rewrite chunk system // Luminol - Configurable region file format
             // Paper start - rewrite chunk system
             if (region == null) {
                 // if the RegionFile doesn't exist, no point in deleting from it
@@ -414,12 +426,12 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
                 DataOutputStream output = region.getChunkDataOutputStream(pos);
                 try { // Paper - Only write if successful
                     NbtIo.write(value, output);
-                    region.setOversized(pos.x(), pos.z(), false); // Paper - We don't do this anymore, mojang stores differently, but clear old meta flag if it exists to get rid of our own meta file once last oversized is gone
+                    if (region instanceof RegionFile mcaFile) mcaFile.setOversized(pos.x(), pos.z(), false); // Paper - We don't do this anymore, mojang stores differently, but clear old meta flag if it exists to get rid of our own meta file once last oversized is gone // Luminol - Configurable region file format
                     // Paper start - don't write garbage data to disk if writing serialization fails
                     output.close();
                 } catch (final RegionFileSizeException ex) {
                     region.clear(pos);
-                    final int maxSize = RegionFile.MAX_CHUNK_SIZE / (1024 * 1024);
+                    final int maxSize = (region instanceof RegionFile ? RegionFile.MAX_CHUNK_SIZE : abomination.LinearRegionFile.MAX_CHUNK_SIZE) / (1024 * 1024); // Luminol - Configurable region file format
                     LOGGER.error("Chunk at (" + pos.x() + "," + pos.z() + ") in regionfile '" + region.getPath().toString() + "' exceeds max size of " + maxSize + "MiB, it has been deleted from disk.");
                     // Paper end - don't write garbage data to disk if writing serialization fails
                 }
@@ -432,7 +444,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         // Paper start - rewrite chunk system
         synchronized (this) {
             final ExceptionCollector<IOException> exceptionCollector = new ExceptionCollector<>();
-            for (final RegionFile regionFile : this.regionCache.values()) {
+            for (final abomination.IRegionFile regionFile : this.regionCache.values()) {
                 try {
                     regionFile.close();
                 } catch (final IOException ex) {
@@ -448,7 +460,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         // Paper start - rewrite chunk system
         synchronized (this) {
             final ExceptionCollector<IOException> exceptionCollector = new ExceptionCollector<>();
-            for (final RegionFile regionFile : this.regionCache.values()) {
+            for (final abomination.IRegionFile regionFile : this.regionCache.values()) {
                 try {
                     regionFile.flush();
                 } catch (final IOException ex) {
